## 1. Project Overview

**Project:** Predicting Rural Water Point Functionality Using Data Mining

**Target variable:** `functional3`

**Task:** Multiclass classification

**Target classes:**
- Functional
- Partially functional
- Abandoned or not functional

This notebook covers the **initial data understanding** step only: loading the raw dataset and exploring its basic structure. No cleaning, preprocessing, feature engineering, or modeling is performed here.

## 2. Dataset Source

**Dataset:** Rural Water Point Functionality Dataset (`rwpfunctionality`)

**Source:** https://openwashdata.github.io/rwpfunctionality/

**File location (raw, unmodified):** `data/raw/water_pump_functionality.xlsx`

## 3. Loading the Dataset

We load the raw Excel file using `pandas`. The path is written **relative** to the `notebooks/` folder so the notebook works on any machine, regardless of where the project is stored.

We only read the file here — it is never modified or overwritten.

In [ ]:
# Import pandas for data loading and exploration
import pandas as pd

# Relative path to the raw dataset (relative to the notebooks/ folder)
DATA_PATH = "../data/raw/water_pump_functionality.xlsx"

# Load the Excel file into a DataFrame
df = pd.read_excel(DATA_PATH)

print("Dataset loaded successfully.")

Dataset loaded successfully.


## 4. Dataset Dimensions

Check how many rows (observations) and columns (variables) the dataset has.

In [ ]:
# df.shape returns a tuple: (number_of_rows, number_of_columns)
n_rows, n_cols = df.shape

print(f"Number of rows (observations): {n_rows}")
print(f"Number of columns (variables): {n_cols}")

Number of rows (observations): 1793
Number of columns (variables): 52


## 5. First Few Records

Look at the first and last few rows to get a feel for the raw data.

In [ ]:
# First 5 rows of the dataset
df.head()

,id,country,admin1,admin2,admin3,instance_wp,subdate_wp,latitude_wp,longitude_wp,elevation_wp,...,qtyhh_c_wp,wc_admin_index_wp,wc_finance_index_wp,wc_mgmt_index_wp,wc_maint_index_wp,wc_savings_wp,wpqty_wp,pop_1000,annual_rain,season
0,1,Ethiopia,Tigray,North Western,Medebay Zana,762331022,2020-01-02,14.116315,38.493830,2033.5,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2,Ethiopia,Tigray,North Western,Tselemti,751461172,2019-12-21,13.581540,38.467207,1428.1,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,3,Ethiopia,Tigray,North Western,Medebay Zana,758401118,2019-12-29,14.097803,38.452013,1997.3,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,4,Ethiopia,Tigray,North Western,Medebay Zana,769951024,2019-12-29,13.912499,38.437931,1877.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,5,Ethiopia,Tigray,North Western,Tahtay Koraro,765940926,2019-12-18,14.041615,38.437492,1966.3,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [ ]:
# Last 5 rows of the dataset
df.tail()

,id,country,admin1,admin2,admin3,instance_wp,subdate_wp,latitude_wp,longitude_wp,elevation_wp,...,qtyhh_c_wp,wc_admin_index_wp,wc_finance_index_wp,wc_mgmt_index_wp,wc_maint_index_wp,wc_savings_wp,wpqty_wp,pop_1000,annual_rain,season
1788,1789,Uganda,Eastern,Namayingo,Mutumba,mwater2350-63U8UZ,2017-10-23,0.267397,33.785696,1163.680309,...,200.0,Minimum,Moderate,Moderate,Moderate,1.0,2.0,2523.028071,1286.0,wet
1789,1790,Uganda,Eastern,Namayingo,Mutumba,mwater2689-63WSHC,2017-10-24,0.262596,33.774656,1141.099976,...,100.0,Moderate,Moderate,Moderate,Moderate,1.0,1.0,2069.105378,1286.0,wet
1790,1791,Uganda,Eastern,Namayingo,Mutumba,mwater2343-63WW64,2017-10-24,0.248204,33.769656,1150.646093,...,150.0,Inadequate,Minimum,Inadequate,Moderate,1.0,3.0,1888.559240,1286.0,wet
1791,1792,Uganda,Eastern,Namayingo,Mutumba,mwater2343-6447R5,2017-10-26,0.210286,33.757012,1140.286588,...,300.0,Inadequate,Minimum,Inadequate,Moderate,1.0,3.0,1422.968472,1286.0,wet
1792,1793,Uganda,Eastern,Namayingo,Mutumba,mwater2686-64468B,2017-10-26,0.230352,33.748044,1143.699951,...,100.0,Inadequate,Moderate,Inadequate,Moderate,1.0,2.0,1304.366727,1286.0,wet


## 6. Dataset Information

`df.info()` gives a compact summary: column names, non-null counts, and data types in one view.

In [ ]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1793 entries, 0 to 1792
Data columns (total 52 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   id                        1793 non-null   int64  
 1   country                   1793 non-null   str    
 2   admin1                    1793 non-null   str    
 3   admin2                    1760 non-null   str    
 4   admin3                    1592 non-null   str    
 5   instance_wp               1793 non-null   str    
 6   subdate_wp                1793 non-null   str    
 7   latitude_wp               1767 non-null   float64
 8   longitude_wp              1767 non-null   float64
 9   elevation_wp              1658 non-null   float64
 10  commid_wp                 1793 non-null   str    
 11  dataorg_wp                1793 non-null   str    
 12  timepoint_wp              1793 non-null   str    
 13  cwfunded_wp               1780 non-null   str    
 14  cwid               

## 7. Column Names

List every column name in the dataset.

In [ ]:
# List of all column names
list(df.columns)

['id',
 'country',
 'admin1',
 'admin2',
 'admin3',
 'instance_wp',
 'subdate_wp',
 'latitude_wp',
 'longitude_wp',
 'elevation_wp',
 'commid_wp',
 'dataorg_wp',
 'timepoint_wp',
 'cwfunded_wp',
 'cwid',
 'wptype',
 'pumptype',
 'drillmethod',
 'piped_source',
 'piped_pump',
 'rehabyn',
 'qtyhh_wp',
 'whomanage_wp',
 'wp_age',
 'rehab_age',
 'qtypeople_wp',
 'wateravailable',
 'functional3',
 'whynowatertoday_wp',
 'whynowatertoday_wp_other',
 'minutesfill20l',
 'brokendays_wp',
 'qtymonthsnowater_wp',
 'downtime2weeks',
 'lockedfullday_wp',
 'pumpstrokes',
 'photo_wp',
 'grant_number_wp',
 'wc_present_wp',
 'paytocollect_wp',
 'balance_any_dollars_wp',
 'improved_wponly_wp',
 'qtyhh_c_wp',
 'wc_admin_index_wp',
 'wc_finance_index_wp',
 'wc_mgmt_index_wp',
 'wc_maint_index_wp',
 'wc_savings_wp',
 'wpqty_wp',
 'pop_1000',
 'annual_rain',
 'season']

## 8. Data Types

Check the data type of each column, and how many columns fall into each type.

In [ ]:
# Data type of each column
df.dtypes

id                            int64
country                         str
admin1                          str
admin2                          str
admin3                          str
instance_wp                     str
subdate_wp                      str
latitude_wp                 float64
longitude_wp                float64
elevation_wp                float64
commid_wp                       str
dataorg_wp                      str
timepoint_wp                    str
cwfunded_wp                     str
cwid                            str
wptype                          str
pumptype                        str
drillmethod                     str
piped_source                    str
piped_pump                      str
rehabyn                         str
qtyhh_wp                    float64
whomanage_wp                    str
wp_age                      float64
rehab_age                   float64
qtypeople_wp                float64
wateravailable                  str
functional3                 

In [ ]:
# Count how many columns share each data type
df.dtypes.value_counts()

str        32
float64    19
int64       1
Name: count, dtype: int64

## 9. Basic Summary Statistics

This section covers basic data-quality checks: descriptive statistics, missing values, duplicate rows, and the number of unique values per column. We are only **observing** these facts — no values are removed or changed.

In [ ]:
# Descriptive statistics for all columns (numeric and categorical)
# include="all" so text/categorical columns are summarized too, not just numeric ones
df.describe(include="all")

,id,country,admin1,admin2,admin3,instance_wp,subdate_wp,latitude_wp,longitude_wp,elevation_wp,...,qtyhh_c_wp,wc_admin_index_wp,wc_finance_index_wp,wc_mgmt_index_wp,wc_maint_index_wp,wc_savings_wp,wpqty_wp,pop_1000,annual_rain,season
count,1793.000000,1793,1793,1760,1592,1793,1793,1767.000000,1767.000000,1658.000000,...,1195.000000,1120,1120,1120,1120,1191.000000,1195.000000,1229.000000,1246.000000,1246
unique,NaN,9,15,30,108,1793,338,NaN,NaN,NaN,...,NaN,4,4,4,4,NaN,NaN,NaN,NaN,2
top,NaN,Ethiopia,Tigray,Janakpur,Sindhuli,762331022,2018-06-29,NaN,NaN,NaN,...,NaN,Moderate,Moderate,Inadequate,Moderate,NaN,NaN,NaN,NaN,dry
freq,NaN,386,386,280,280,1,19,NaN,NaN,NaN,...,NaN,598,648,443,519,NaN,NaN,NaN,NaN,786
mean,897.000000,NaN,NaN,NaN,NaN,NaN,NaN,8.664635,42.510953,929.635329,...,238.751464,NaN,NaN,NaN,NaN,72.679261,5.791632,931.880195,996.455056,NaN
std,517.738834,NaN,NaN,NaN,NaN,NaN,NaN,15.079995,29.163977,774.583928,...,306.556354,NaN,NaN,NaN,NaN,257.849626,5.820825,1041.590118,395.776235,NaN
min,1.000000,NaN,NaN,NaN,NaN,NaN,NaN,-17.634846,-8.595394,-217.000000,...,10.000000,NaN,NaN,NaN,NaN,0.000000,1.000000,31.781268,601.000000,NaN
25%,449.000000,NaN,NaN,NaN,NaN,NaN,NaN,-1.770936,33.473929,303.500000,...,65.000000,NaN,NaN,NaN,NaN,1.000000,2.000000,259.356612,601.000000,NaN
50%,897.000000,NaN,NaN,NaN,NaN,NaN,NaN,13.379742,36.238084,684.721900,...,150.000000,NaN,NaN,NaN,NaN,1.000000,4.000000,568.342223,1007.000000,NaN
75%,1345.000000,NaN,NaN,NaN,NaN,NaN,NaN,21.722116,83.096485,1613.633984,...,349.000000,NaN,NaN,NaN,NaN,1.000000,7.000000,1316.672291,1286.000000,NaN


In [ ]:
# Count missing values per column
missing_counts = df.isnull().sum()

# Keep only columns that actually have missing values, sorted from most to least
missing_summary = missing_counts[missing_counts > 0].sort_values(ascending=False)

# Also express missing values as a percentage of total rows
missing_percent = (missing_summary / len(df) * 100).round(2)

missing_summary_table = pd.DataFrame({
    "missing_count": missing_summary,
    "missing_percent": missing_percent
})

print(f"Columns with at least one missing value: {len(missing_summary_table)} out of {df.shape[1]}")
missing_summary_table

Columns with at least one missing value: 40 out of 52


,missing_count,missing_percent
whynowatertoday_wp_other,1775,99.00
rehab_age,1513,84.38
whynowatertoday_wp,1491,83.16
grant_number_wp,1246,69.49
piped_source,1219,67.99
piped_pump,1219,67.99
brokendays_wp,1160,64.70
balance_any_dollars_wp,853,47.57
cwid,834,46.51
pumpstrokes,818,45.62


In [ ]:
# Count fully duplicated rows (every column identical)
n_duplicates = df.duplicated().sum()

print(f"Number of duplicated rows: {n_duplicates}")

Number of duplicated rows: 0


In [ ]:
# Number of unique values for each column
# Useful for spotting identifier columns (very high uniqueness) or
# near-constant columns (very low uniqueness)
df.nunique().sort_values(ascending=False)

id                          1793
instance_wp                 1793
photo_wp                    1792
longitude_wp                1766
latitude_wp                 1764
commid_wp                   1395
elevation_wp                1387
pop_1000                    1159
cwid                         948
minutesfill20l               625
qtypeople_wp                 472
balance_any_dollars_wp       408
subdate_wp                   338
qtyhh_c_wp                   303
qtyhh_wp                     181
admin3                       108
brokendays_wp                 55
wp_age                        44
pumpstrokes                   39
admin2                        30
wpqty_wp                      29
admin1                        15
qtymonthsnowater_wp           13
annual_rain                   13
whynowatertoday_wp_other      11
rehab_age                     11
whomanage_wp                  11
grant_number_wp               10
dataorg_wp                    10
country                        9
wptype    

## 10. Target Variable Check

Confirm that the expected target column, `functional3`, exists, and look at how the target classes are distributed.

In [ ]:
# Confirm the target column exists before using it
TARGET = "functional3"

if TARGET in df.columns:
    print(f"Target column '{TARGET}' found.")
else:
    print(f"Target column '{TARGET}' NOT found in the dataset.")

Target column 'functional3' found.


In [ ]:
# Count of records in each target class
df[TARGET].value_counts()

functional3
Functional                     1333
Abandoned or not functional     303
Partially functional            157
Name: count, dtype: int64

In [ ]:
# Percentage distribution of each target class
(df[TARGET].value_counts(normalize=True) * 100).round(2)

functional3
Functional                     74.34
Abandoned or not functional    16.90
Partially functional            8.76
Name: proportion, dtype: float64

## Initial Observations

These are purely factual observations from the raw dataset. No preprocessing decisions are made at this stage.

- **Number of observations (rows):** 1,793
- **Number of variables (columns):** 52
- **Data types present:** 32 text/categorical (`str`) columns, 19 `float64` columns, 1 `int64` column (`id`)
- **Missing values:** Present. 40 out of 52 columns have at least one missing value. The most affected columns are `whynowatertoday_wp_other` (1,775 missing), `rehab_age` (1,513 missing), and `whynowatertoday_wp` (1,491 missing).
- **Duplicate rows:** None found (0 fully duplicated rows).
- **Target variable:** `functional3` is present in the dataset, with **3 classes**.
- **Target class counts:**
  - Functional: 1,333
  - Abandoned or not functional: 303
  - Partially functional: 157
- **Target class percentages:**
  - Functional: 74.34%
  - Abandoned or not functional: 16.90%
  - Partially functional: 8.76%
- **Structural characteristics:** The target classes are imbalanced, with "Functional" making up nearly three-quarters of records and "Partially functional" being the smallest class. The dataset contains a mix of identifier/location columns (`id`, `country`, `admin1-3`, `latitude_wp`, `longitude_wp`), water-point attributes (`wptype`, `pumptype`, `wp_age`), and community/management indices (`wc_admin_index_wp`, `wc_finance_index_wp`, etc.), several of which have substantial missingness.

No decisions about removing, imputing, or transforming any variable have been made. These will be addressed in later preprocessing steps.